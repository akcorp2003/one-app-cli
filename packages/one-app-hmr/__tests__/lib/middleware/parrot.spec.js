/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import express from 'express';

import chokidar from 'chokidar';
import {
  loadScenarios, getModuleNameFromFilePath, createHotParrotMiddleware, loadParrotMiddleware,
} from '../../../lib/middleware/parrot';

afterEach(() => {
  jest.clearAllMocks();
});
jest.disableAutomock();
jest.mock('jest-resolve');
jest.mock('express');
jest.mock('parrot-middleware');
const mockExpressRouter = {
  use: jest.fn(() => mockExpressRouter),
  get: jest.fn(() => mockExpressRouter),
  stack: ['test'],
};
express.Router.mockImplementation(() => mockExpressRouter);
jest.mock('chokidar', () => {
  const mockChokidar = {
    watch: jest.fn(() => mockChokidar),
    on: jest.fn(() => mockChokidar),
  };

  return mockChokidar;
});
describe('hot reloads parrot scenarios', () => {
  beforeEach(() => {
    console.log.mockClear();
    console.warn.mockClear();
    console.time.mockClear();
  });
  const consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation();
  const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation();
  const consoleTimeSpy = jest.spyOn(console, 'time').mockImplementation();
  const scenariosMock = ['./logs'];
  const publish = () => jest.fn();
  it('loading scenarios ', () => {
    const loadScenariosMock = loadScenarios(scenariosMock);
    expect(loadScenariosMock.libName).toEqual('one-app-hmr');
  });
  it('get module name from file path', () => {
    expect(getModuleNameFromFilePath('sample-module/mock/scenarios.js')).toEqual('sample-module');
  });
  it('on error log the error', () => {
    createHotParrotMiddleware(scenariosMock, publish);
    chokidar.on.mock.calls[0][1]();
    expect(consoleWarnSpy.mock.calls).toMatchSnapshot();
    expect(chokidar.on.mock.calls[0][0]).toBe('error');
  });
  test('on ready watches scenarios', () => {
    createHotParrotMiddleware(scenariosMock, publish);
    chokidar.on.mock.calls[1][1]();
    expect(consoleLogSpy.mock.calls).toMatchSnapshot();
    expect(chokidar.on.mock.calls[1][0]).toBe('ready');
  });
  test('on add hot reload scenarios ', () => {
    createHotParrotMiddleware(scenariosMock, publish);
    chokidar.on.mock.calls[2][1]('sample-module/mock/scenarios.js');
    expect(consoleLogSpy.mock.calls).toMatchSnapshot();
    expect(chokidar.on.mock.calls[2][0]).toBe('add');
  });
  test('on change hot reload scenarios ', () => {
    createHotParrotMiddleware(scenariosMock, publish);
    chokidar.on.mock.calls[3][1]('sample-module/mock/scenarios.js');
    expect(consoleLogSpy.mock.calls).toMatchSnapshot();
    expect(chokidar.on.mock.calls[3][0]).toBe('change');
  });
  test('on unlink hot reload scenarios', () => {
    createHotParrotMiddleware(scenariosMock, publish);
    chokidar.on.mock.calls[4][1]('sample-module/mock/scenarios.js');
    expect(consoleLogSpy.mock.calls).toMatchSnapshot();
    expect(chokidar.on.mock.calls[4][0]).toBe('unlink');
  });
  test('loadParrotMiddleware are loaded', async () => {
    const scenarios = ['scenarios.js'];
    const useParrotMiddleware = true;
    await loadParrotMiddleware(mockExpressRouter, {
      scenarios,
      useParrotMiddleware,
      publish,
    });
    expect(chokidar.on.mock.calls[2][0]).toBe('add');
    expect(consoleTimeSpy.mock.calls).toMatchSnapshot();
  });
  test('parrot scenarios are not loaded', async () => {
    const parameters = {
      languagePacks: [],
      useLanguagePacks: true,
      publish,
    };

    await loadParrotMiddleware(mockExpressRouter, parameters);
    expect(consoleTimeSpy.mock.calls.length).toEqual(0);
  });
});
