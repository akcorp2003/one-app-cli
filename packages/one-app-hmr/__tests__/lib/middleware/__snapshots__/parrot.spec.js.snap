// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hot reloads parrot scenarios loadParrotMiddleware are loaded 1`] = `
Array [
  Array [
    "[38;2;255;228;196m[1m  one-app-hmr ::[22m[39m [38;2;255;20;147mparrot[39m - initializing",
  ],
]
`;

exports[`hot reloads parrot scenarios on add hot reload scenarios  1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;255;20;147mparrot[39m - Scenarios for [38;2;255;165;0m\\"sample-module\\"[39m has been added",
  ],
]
`;

exports[`hot reloads parrot scenarios on change hot reload scenarios  1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;255;20;147mparrot[39m - Scenarios for \\"sample-module\\" has been changed",
  ],
]
`;

exports[`hot reloads parrot scenarios on error log the error 1`] = `
Array [
  Array [
    "
[38;2;255;165;0m[1m  one-app-hmr ::[22m[39m [33m(Warning)[39m [38;2;255;165;0m[38;2;255;20;147mparrot[39m[38;2;255;165;0m - Watch error: undefined[39m
",
  ],
]
`;

exports[`hot reloads parrot scenarios on ready watches scenarios 1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;255;20;147mparrot[39m - Watching scenarios",
  ],
]
`;

exports[`hot reloads parrot scenarios on unlink hot reload scenarios 1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;255;20;147mparrot[39m - Scenarios for \\"sample-module\\" has been removed",
  ],
]
`;
